File: _CodeTopics/LeetCode/401-600/000539/000539.py
class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:

        length = len(timePoints)
        if len(set(timePoints)) < length:
            return 0
        
        def calculate_diff(t1, t2):
            hours = int(t2[:2]) - int(t1[:2])
            minutes = int(t2[3:]) - int(t1[3:])
            minutes += hours * 60
            return min(minutes, 60*24 - minutes)
        
        timePoints.sort()
        # 这里按常理说应该是 mindiff = calculate_diff(timePoints[-1], timePoints[0]) 的，
        # 因为要第一个减去最后一个，但是 calculate_diff 假定了第二个入参更大。
        # 其实只要代入测试用例 ["23:59","00:00"] 就会得到 -1439，就知道为啥错了。
        mindiff = calculate_diff(timePoints[0], timePoints[-1])
        for i in range(1, length):
            mindiff = min(mindiff, calculate_diff(timePoints[i-1], timePoints[i]))
        return mindiff
        
"""
https://leetcode-cn.com/submissions/detail/259632531/

执行用时：32 ms, 在所有 Python3 提交中击败了93.10%的用户
内存消耗：16.5 MB, 在所有 Python3 提交中击败了63.60%的用户
通过测试用例：
113 / 113
"""

File: _CodeTopics/LeetCode/401-600/000539/tran/000539.cpp
class Solution {
public:
    int findMinDifference(vector<string>& timePoints) {
        
        int length = timePoints.size();
        unordered_set<string> se(timePoints.begin(), timePoints.end());
        if (se.size() < length) {
            return 0;
        }

        std::function<int(string, string)> calculate_diff = [](string t1, string t2) {
            int hours = stoi(t2.substr(0, 2)) - stoi(t1.substr(0, 2));
            int minutes = stoi(t2.substr(3, 2)) - stoi(t1.substr(3, 2));
            minutes += hours * 60;
            return min(minutes, 60*24 - minutes);
        };

        std::sort(timePoints.begin(), timePoints.end());
        int mindiff = calculate_diff(timePoints[0], timePoints.back());
        for (int i = 1; i < length; ++i) {
            mindiff = min(mindiff, calculate_diff(timePoints[i-1], timePoints[i]));
        }
        return mindiff;
    }
};

/*
https://leetcode-cn.com/submissions/detail/259634329/

执行用时：8 ms, 在所有 C++ 提交中击败了91.06%的用户
内存消耗：13.7 MB, 在所有 C++ 提交中击败了9.12%的用户
通过测试用例：
113 / 113
*/
